oauth/lib/redirectlistener.cpp

```diff
@@ -0,0 +1,325 @@
+  "<!DOCTYPE html>\\r\\n"
+  "<html></html>\\r\\n\\r\\n"
+  \endcode
+*/
+QString SailfishOS::OAuth::RedirectListener::httpContent() const
+{
+    Q_D(const RedirectListener);
+    return d->m_httpContent;
+}
+
+void SailfishOS::OAuth::RedirectListener::setHttpContent(const QString &httpContent)
+{
+    Q_D(RedirectListener);
+    if (d->m_httpContent != httpContent) {
+        d->m_httpContent = httpContent;
+    }
+}
+
+/*!
+  \brief Stop listening for redirects from the system web browser.
+
+  This method will be called automatically if the \l{timeout} interval is
+  reached when listening for directs, or once a valid redirect is captured.
+
+  Clients may call this manually in order to stop listening.
+*/
+void SailfishOS::OAuth::RedirectListener::stopListening()
+{
+    Q_D(RedirectListener);
+    if (d->m_tcpServer) {
+        d->m_tcpServer->close();
+        d->m_tcpServer->deleteLater(); // will delete m_timeoutTimer (child).
+        d->m_tcpServer = nullptr;
+        d->m_timeoutTimer = nullptr;
+    }
+}
+
+/*!
+  \brief Start listening for redirects from the system web browser.
+
+  The client application should call this method in order to make the helper
+  start to listen for redirects.  If a \l{port} has been specified by the
+  client application, the helper will attempt to listen on that port;
+  otherwise, the helper will attempt to listen on an arbitrary port above 1024.
+
+  If the helper is unable to listen (e.g. due to the specified \l{port} being
+  unavailable, or due to no ports being available to the application if the
+  \l{port} is left as the default 0 value), it will \l{stopListening()} and
+  emit the \l{failed()} signal.
+
+  Otherwise, the helper will listen for up to \l{timeout} seconds before
+  emitting the \l{timedOut()} signal and \l{stopListening()}.
+
+  If a redirect is received, it will write the \l{httpContent} to the socket,
+  and then \l{stopListening()} before emitting \l{receivedRedirect()}.
+*/
+void SailfishOS::OAuth::RedirectListener::startListening()
+{
+    Q_D(RedirectListener);
+    if (d->m_tcpServer) {
+        // already listening
+        return;
+    }
+
+    d->m_tcpServer = new QTcpServer(this);
+    d->m_timeoutTimer = new QTimer(d->m_tcpServer);
+    d->m_timeoutTimer->setInterval(d->m_timeout * 1000);
+    d->m_timeoutTimer->setSingleShot(true);
+    d->m_timeoutTimer->start();
+    connect(d->m_timeoutTimer, &QTimer::timeout,
+            d->m_timeoutTimer, [this] {
+        stopListening();
+        emit timeout();
+    });
+
+    bool found = false;
+    if (d->m_port > 0 && d->m_port < 65534) {
+        // use the port specified by the client
+        const quint16 p = d->m_port;
+        if (d->m_tcpServer->listen(QHostAddress(QHostAddress::LocalHost), p)) {
+            d->m_uri = QStringLiteral("http://127.0.0.1:%1/").arg(p);
+            emit uriChanged();
+            found = true;
+        }
+    } else {
+        // search for an available port
```

Comment:
I think you should not do it. Port is a part of a contract, that you specify when you register a client

Reply:
Do you mean, on the server side?  The few services I've interacted with haven't required the client application to define the port they listen on, if using a localhost redirect.

Or do you simply mean, it's kind of bad to just search for an available port, in general?  That's fair enough, I suppose.  I'd like to keep the ability for the user to not care about redirect at all, in the common case (e.g. using OAuth2AcPkce.qml type), so I wonder what the best way is (e.g. I'd prefer not to expose a "redirectPort" property at the top level, so better might be to just specify a default port to use, although this might cause issues if that port is used by something else or if multiple applications are running, etc)...

Reply:
I meant that even if you find a free port and open a socket on it, the server will still make the last redirect to port that you specify when registered the client on the server. There may be a situation when the original port is busy, then you found another one, but no redirect from the browser or webview will come, because the server does not know about the port been changed. For example, the authentication service that we use will return an error if you pass (in first authentication request) it a redirect uri different from the registered one.
First point in https://www.ory.sh/hydra/docs/debug/
Maybe there are existing servers with more relaxed requirements, that allow changing redirect uri. In that case, perhaps you can add some kind of option (searchAvailablePort(true | false)) that allows to switch this behaviour.

Reply:
I see. Google allows to change port
https://developers.google.com/identity/protocols/oauth2/native-app#step-2:-send-a-request-to-googles-oauth-2.0-server

Reply:
Yeah.  I think having the ability to "automatically" choose a port is useful.  The client can set the port manually if required (e.g. `redirectListener.port: 12345`) in which case the RedirectListener will use that port only.  I've updated the OAuth2AcPkce documentation and example to set the port manually, as it's likely to be a common use-case.